extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java

```diff
@@ -87,12 +81,27 @@ public Object get(ByteBuffer buf, int position)
   //Note that this is not threadsafe and I don't think it needs to be
   private Union getUnion(ByteBuffer buf, int position)
   {
-    Union union = unions.get(position);
-    if (union == null) {
-      Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
-      union = (Union) SetOperation.wrap(mem);
-      unions.put(position, union);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
+    Union union = unionMap != null ? unionMap.get(position) : null;
+    if (union != null) {
+      return union;
+    }
+    return createNewUnion(buf, position, true);
+  }
+
+  private Union createNewUnion(ByteBuffer buf, int position, boolean isWrapped)
+  {
+    NativeMemory nm = getNativeMemory(buf);
+    Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
+    Union union = isWrapped
+                  ? (Union) SetOperation.wrap(mem)
+                  : (Union) SetOperation.builder().initMemory(mem).build(size, Family.UNION);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
```

Comment:
@gianm Yes, `SketchAggregator.updateUnion` is expected to take longer than `getUnion` and requested optimization might not yield performance difference.
If BufferAggregators called in the order you mentioned in https://github.com/druid-io/druid/pull/4071#discussion_r107329220, then `nmCache` and `unions` will contain only one key.
>whether or not this optimization is necessary right now.

I think the current changes are in good state, no optimization required and are ready to be merged.

>However instead of addressing my original small code quality comment (from last week) that would have taken less than 5 min, @akashdw started a dispute and then a super time-draining benchmark battle

@leventov Are you suggesting to accept change request without doing validation? 

To be very clear, one can not simply accept these kind of change requests without verification and validation b/c its about performance and not about changes you requested which can be done in 5 minutes.

fyi your initial recommendation had bug (b/c probably you never validated or benchmarked it )and caught during benchmarking. 
Given,`SketchAggregator.updateUnion` is expected to take longer than `getUnion` and requested optimization might not yield performance difference, can you please unblock this PR ?